### PR TITLE
BFloat16: add explicit dtype support for to_mkldnn and to_dense

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37218 BFloat16: enable prepacked weights's inference
* **#37215 BFloat16: add explicit dtype support for to_mkldnn and to_dense**
* #37199 add mkldnn relu backward and reshape backward
* #37147 add mkldnn batch_norm backward
* #37146 add mkldnn pooling backward
* #36122 add mkldnn linear backward
* #36121 enable mkldnn conv2d backward to support mkldnn tensor input

Differential Revision: [D22440964](https://our.internmc.facebook.com/intern/diff/D22440964)

### CODE REVIEW

**Comment by XiaobingSuper** [comment]
@jgong5, @hongzhen1

File: aten/src/ATen/native/sparse/SparseTensor.cpp [hunk 1]
```diff
@@ -337,7 +337,7 @@ SparseTensor dense_to_sparse(const Tensor& self, int64_t sparse_dim){
 
 // NB: Dropped the resizeNd variants
 
-Tensor sparse_to_dense(const SparseTensor& self) {
+Tensor sparse_to_dense(const SparseTensor& self, c10::optional<ScalarType> dtype) {
```
Comment (albanD) [inline aten/src/ATen/native/sparse/SparseTensor.cpp:340]: The new flag is completely ignored for here. Is that expected?

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 2]
```diff
@@ -8,35 +8,53 @@ namespace at { namespace native {
...
-  AT_ASSERTM(cpu_tensor.device().type() == DeviceType::CPU,
-             "dense_to_mkldnn expects CPU tensor input");
-  AT_ASSERTM(cpu_tensor.layout() == Layout::Strided,
-             "dense_to_mkldnn expects strided tensor input");
-  AT_ASSERTM(cpu_tensor.scalar_type() == ScalarType::Float,
-             "dense_to_mkldnn expects float tensor input");
-  AT_ASSERTM(cpu_tensor.dim() <= 5,
-             "Can't convert cpu tensor with the number of dimensions > 5");
+Tensor dense_to_mkldnn(const Tensor& cpu_tensor, c10::optional<ScalarType> dtype) {
+  TORCH_INTERNAL_ASSERT(cpu_tensor.device().type() == DeviceType::CPU,
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:34]: Why are these internal asserts? These args are not controled by the user? If so, you prefer a TORCH_CHECK here I think.

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 3]
```diff
@@ -8,35 +8,53 @@ namespace at { namespace native {
...
-  Tensor mkldnn_tensor = empty_mkldnn(cpu_tensor_cont.sizes(), cpu_tensor_cont.options());
+  auto data_type = dtype.has_value() ? dtype.value() : cpu_tensor.scalar_type();
+  Tensor mkldnn_tensor = empty_mkldnn(
+      cpu_tensor_cont.sizes(), cpu_tensor_cont.options().dtype(data_type));
   ideep::tensor& dtensor = itensor_from_mkldnn(mkldnn_tensor);
-  dtensor.feed_from(dtensor.get_dims(),
-                    ideep::tensor::data_type::f32,
-                    (cpu_tensor_cont.template data_ptr<float>()));
+  if (cpu_tensor.scalar_type() == ScalarType::Float) {
+    dtensor.feed_from(dtensor.get_dims(),
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:50]: This function will work for any `dtype` provided?

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 4]
```diff
@@ -8,35 +8,53 @@ namespace at { namespace native {
...
   // NOTE: int32_t dims from ideep::tensor but sizes needs int64_t
+  auto data_type = dtype.has_value() ? dtype.value() : mkldnn_tensor.scalar_type();
   Tensor cpu_tensor = at::empty(
     std::vector<int64_t>(dims.begin(), dims.end()),
-    mkldnn_tensor.options().layout(c10::kStrided));
+    mkldnn_tensor.options().layout(c10::kStrided).dtype(data_type));
   if (stensor.is_empty()) return cpu_tensor;
-  auto pub_tensor = stensor.to_public(cpu_tensor.template data_ptr<float>());
+  auto pub_tensor =
+      cpu_tensor.scalar_type() == ScalarType::Float
```
Comment (albanD) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:24]: Here the cpu_tensor type is dtype. So it can be other things that float/bfloat right?

**Review by albanD — CHANGES_REQUESTED** [review]
How general is this supposed to be? If there are restrictions on the dtype, a couple checks are missing I think.

Also for sparse Tensor, if the output dtype arg is not supported, and error should be raised if the user passes it.

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp (hunk 3 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:50]: Yes, the default dtype is float, there can support othe dtype, such as int8.

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 1 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/sparse/SparseTensor.cpp:340]: This is just for dispatch which should have same numbers of inputs.

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp (hunk 4 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp:24]: yes.

File: aten/src/ATen/native/sparse/SparseTensor.cpp (hunk 1 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/sparse/SparseTensor.cpp:341]: In this case please error out if dtype is provided, as we prefer not to have situations when user passing ignored argument.

**Review by VitalyFedyunin — COMMENTED** [review]
Several error messaging fixes required, everything else looks good.

**Comment by VitalyFedyunin** [comment]
Please rebase entire stack to make it mergable.

**Pull request closed without merging**